### PR TITLE
Feature/new-hash-formula

### PR DESCRIPTION
**Pull request by janek26**
create upgrade flow to allow old accounts to upgrade to the new tx hash signing.

- [x] copy changes
- [x] update contracts

https://user-images.githubusercontent.com/6831124/164215198-7ac37d08-362e-48e8-996f-41c82af73482.mp4

### CODE REVIEW

File: packages/extension/src/background/upgrade.ts [hunk 1]
```diff
@@ -0,0 +1,49 @@
+import {
+  AddTransactionResponse,
+  KeyPair,
+  ProviderInterface,
+  number,
+} from "starknet"
+import { Account as OldAccountV390, stark as starkV390 } from "starknet-390"
```
Comment (delaaxe) [inline packages/extension/src/background/upgrade.ts:7]: either `OldAccount` or `AccountV390` should be enough

File: packages/extension/src/background/upgrade.ts [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+
+        return oldAccount.execute(
+          {
+            contractAddress: accountAddress,
+            entrypoint: "upgrade",
+            calldata: starkV390.compileCalldata({
+              implementation: newImplementation,
+            }),
+          },
+          undefined,
```
Comment (delaaxe) [inline packages/extension/src/background/upgrade.ts:40]: Unrelated, but the `starknet.js` DX should really be improved, this is error prone

File: packages/extension/src/background/wallet.ts [hunk 3]
```diff
@@ -172,24 +172,49 @@ export class Wallet extends EventEmitter {
...
+  public async getStarknetAccountByAddress(address: string): Promise<Account> {
     if (!this.isSessionOpen()) {
       throw Error("no open session")
     }
-
-    const account = await this.getSelectedAccount()
+    const account = this.getAccountByAddress(address)
     if (!account) {
-      throw new Error("no selected account")
+      throw Error("account not found")
```
Comment (delaaxe) [inline packages/extension/src/background/wallet.ts:195]: image

File: packages/extension/src/background/index.ts [hunk 4]
```diff
@@ -247,6 +235,43 @@ const successStatuses = ["ACCEPTED_ON_L1", "ACCEPTED_ON_L2", "PENDING"]
...
+          starknetAccount, // Account extends Provider
+          wallet.getKeyPairByDerivationPath(account.signer.derivationPath), // signer is a private property of the account, this will be public in the future
+        )
+
+        return transactionTracker.trackTransaction(
+          updateTransaction.transaction_hash,
+          account,
+          {
+            title: "Upgrading account",
+          },
```
Comment (delaaxe) [inline packages/extension/src/background/index.ts:271]:
```suggestion
          { title: "Upgrading account" },
```

File: packages/extension/src/background/upgrade.ts [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+  keyPair: KeyPair,
+) => Promise<AddTransactionResponse>) => {
+  switch (true) {
+    case number
+      .toBN(oldImplementation)
+      .eq(
+        number.toBN(
+          "0x0090aa7a9203bff78bfb24f0753c180a33d4bad95b1f4f510b36b00993815704",
+        ),
+      ):
```
Comment (delaaxe) [inline packages/extension/src/background/upgrade.ts:24]: could be improved

**Review by delaaxe — APPROVED** [review]

File: packages/extension/src/background/wallet.ts (hunk 3 above)
Comment (janek26) [reply packages/extension/src/background/wallet.ts:195]: Bildschirmfoto 2022-04-20 um 14 34 10

something to search and replace in a separate PR

File: packages/extension/src/background/upgrade.ts [hunk 6]
```diff
@@ -0,0 +1,50 @@
...
+  keyPair: KeyPair,
+) => Promise<AddTransactionResponse>) => {
+  if (
+    number
+      .toBN(oldImplementation)
+      .eq(
+        number.toBN(
+          "0x0090aa7a9203bff78bfb24f0753c180a33d4bad95b1f4f510b36b00993815704",
+        ),
+      )
```
Comment (delaaxe) [inline packages/extension/src/background/upgrade.ts:24]: Cleaner to use one or two temporary variables here

Comment (janek26) [reply packages/extension/src/background/upgrade.ts]: done! thanks!

**Pull request merged**